server/services/transaction.service.js

```diff
@@ -73,16 +74,56 @@ export const getPaymentDuration = async (offerId) => {
   return { isLastPayment, isFirstPayment, isOtherPayment };
 };
 
+export const processOfferTransaction = async ({ model, newTransaction }) => {
+  await generateNextPaymentDate({
+    transactionId: newTransaction._id,
+    offerId: model._id,
+  }).catch((error) => {
+    throw new ErrorHandler(httpStatus.INTERNAL_SERVER_ERROR, 'Internal Server Error', error);
+  });
+
+  const paymentType = await getPaymentDuration(model._id);
+
+  if (paymentType.isLastPayment) {
+    const isFirstSale = await isVendorFirstSale(model.vendorId);
+
+    if (isFirstSale) {
+      await assignBadgeAutomatically(AUTOMATED_BADGES.VENDOR_FIRST_SALE, model.vendorId);
+    }
+  }
+
+  const referral = await getReferralByOfferId(model._id);
+
+  if (referral) {
+    await updateReferralAccumulatedRewardAndRewardStatus({
+      referralId: referral._id,
+      transactionId: newTransaction._id,
+      amountPaid: newTransaction.amount,
+      offerId: model._id,
+    });
+  }
+};
+
 export const addTransaction = async (transaction) => {
+  let offerTransactionParameters = {};
```

Comment:
This is better as transaction parameters to ensure that it can be reused by other models.
```suggestion
let transactionParameters = {};
```